handler/src/main/java/io/netty5/handler/ssl/EngineWrapper.java

```diff
@@ -0,0 +1,222 @@
+                assert count > 0: "Input buffer has readable bytes, but no readable components: " + in;
+                inputs = count == 1 ? singleReadableBuffer : new ByteBuffer[count];
+                int prepared = in.forEachReadable(0, this);
+                assert prepared == count: "Expected to prepare " + count + " buffers, but got " + prepared;
+            } else {
+                inputs = singleReadableBuffer;
+                int readable = in.readableBytes();
+                if (cachedReadingBuffer == null || cachedReadingBuffer.capacity() < readable) {
+                    cachedReadingBuffer = allocateCachingBuffer(readable);
+                }
+                cachedReadingBuffer.clear();
+                in.copyInto(in.readerOffset(), cachedReadingBuffer, 0, readable);
+                cachedReadingBuffer.limit(readable);
+                inputs[0] = cachedReadingBuffer;
+            }
+        }
+        if (out == null || out.writableBytes() == 0) {
+            outputs = singleEmptyBuffer;
+        } else {
+            if (out.isDirect() == useDirectBuffer) {
+                int count = out.countWritableComponents();
+                assert count > 0: "Output buffer has writable space, but no writable components: " + out;
+                outputs = count == 1 ? singleWritableBuffer : new ByteBuffer[count];
+                int prepared = out.forEachWritable(0, this);
+                assert prepared == count: "Expected to prepare " + count + " buffers, but got " + prepared;
+            } else {
+                inputs = singleWritableBuffer;
+                int writable = out.writableBytes();
+                if (cachedWritingBuffer == null || cachedWritingBuffer.capacity() < writable) {
+                    cachedWritingBuffer = allocateCachingBuffer(writable);
+                }
+                outputs[0] = cachedWritingBuffer.position(0).limit(writable);
+                writeBack = true;
+            }
+        }
+    }
+
+    private ByteBuffer allocateCachingBuffer(int capacity) {
+        capacity = PlatformDependent.roundToPowerOfTwo(capacity);
+        return useDirectBuffer ? ByteBuffer.allocateDirect(capacity) : ByteBuffer.allocate(capacity);
+    }
+
+    private void limitInput(int length) {
+        for (ByteBuffer input : inputs) {
+            int remaining = input.remaining();
+            if (remaining > length) {
+                input.limit(input.position() + length);
+                length = 0;
+            } else {
+                length -= remaining;
+            }
+        }
+    }
+
+    private SSLEngineResult processResult(SSLEngineResult result) {
+        this.result = result;
+        return result;
+    }
+
+    private void finish(Buffer in, Buffer out) {
+        if (result != null) { // Result can be null if the operation failed.
+            if (in != null) {
+                in.skipReadable(result.bytesConsumed());
+                Reference.reachabilityFence(in);
```

Comment:
please add comments why we use a `reachableFence(...)`. The same comment holds true for below.

Reply:
Yes, this also needs to be restructured a bit.